lib/token.h

```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
     explicit Token(TokensFrontBack *tokensFrontBack = nullptr);
     ~Token();
 
+    struct TokenEnumeration
+    {
+        Token const* mt;
+        TokenEnumeration(Token const* t) : mt(t) {}
+
+        struct TokenIterator
```

Comment:
For compatibility with C++11 algorithms, you should add these typedefs:

```cpp
using iterator_category = std::forward_iterator_tag;
using value_type = Token*;
using difference_type = std::ptrdiff_t;
using pointer = Token*;
using reference = Token*;
```

Reply:
it looks wrong to me that value_type, pointer and reference would all be the same type. Shouldn't pointer be either Token** or void?

Reply:
The `reference` type should be the result of `operator*` and `pointer` should be the result of `operator->`. You are missing `operator->` in your iterator, but depending on how you define `->` this typedef should be set accordingly.

Reply:
How would you define ```->```? Normally I would have it return a pointer to the thing being iterated over which in this case would be ```T**```. That  doesn't seem very useful though. You could make it ```T*``` instead but that would mean effectively a double dereference which seems bad as well. I think the best option might be to make pointer ```void``` to reflect the fact that there is no operator->

Reply:
It would be useful to have a `->` especially when used with algorithms. Something like:

```cpp
...
auto it = std::find_if(start, last, predicate);
if (it != last) {
    return it->values();
}
...
```

For such an example to work we would need to return `T*`, I believe. If we return `T**` it would require an extra dereference to work(`operator->` will only drill down until it hits a pointer type, which means an extra dereference is necessary). Either way, I dont think its possible to return `T**` as it would create a dangling pointer. To avoid this, you could return proxy instead:

```cpp
        struct proxy {
            T* x;
            T* operator->() const { return x; }
        };
        proxy operator->() const { return proxy{mt}; }
```

But it seems like just returning `T*` is simpler.

Reply:
Of course, you could leave it as `void` for now, and we can add `operator->` at a later time.

Reply:
I agree T* makes more sense than T** or proxy. My concern is that it->values makes it look like it as a Token* which it isn't.